### PR TITLE
Set up vault client and deposit/redeem handling

### LINKED ISSUE #5: Set up vault client and check for deposits
The vault client is a service that runs independently from a blockchain node.

Every vault has an account on Stellar, it's vault account. The secret key of the vault account is private to the vault service and not shared with anyone.

Furthermore, every vault has a reference to a substrate node that has the Spacewalk pallet.

The vault should regularly check Stellar for new payment transactions to its vault account (regularly = every 5 seconds; and check = poll the server for transactions). If there is such a transaction, it will then call the `report_stellar_transaction` extrinsic of the Spacewalk pallet of the connected Substrate node to report this transaction.

### LINKED ISSUE #6: Vault: monitor redeems in the bridge pallet and unlock funds
The vault will monitor the substrate chain it is linked to and check whether it successfully executed a transaction that called the extrinsic `initiate_redeem` (this needs to be a finalized transaction).

It will then check whether the vault address specified in the transaction is targeted at the vault reading this transaction. If that is the case, then the vault will submit a Stellar transaction to unlock the funds from its vault account and send it to the users account.

This operation will be the same as [`execute_withdrawal`](https://github.com/pendulum-chain/pendulum-prototype/blob/bridge-poc/pallets/stellar-bridge/src/lib.rs#L336) of the pendulum prototype.

### LINKED ISSUE #26: The Native currency given by the faucet is not the same required to pay fees
Note that on [this guide](https://www.notion.so/satoshipay/Testing-the-Vault-and-Spacewalk-9bbdb038fbf344acaccb8aa97ca8e81a#eb43d7052f9f41d99901365a2053ca5a) the step 4. is required to transfer some Native currency to the testing account.
The testing account already have Native currency, but apparently is the one hold by the `pallet_balances`. Since we also have another Native currency defined in our `Currency` definition, I think they might be conflicting here.
We should only rely on `pallet_balances` for managing the native currency (PEN).

### PR DESCRIPTION
**Pull request by ebma**
Adds a vault client that can handle deposits for spacewalk chains.
The vault and used libraries (runtime, service, bitcoin) were copied from the `interbtc-clients` repository.

- [x] Copy vault and make it compile/run
- [x] Make client repeatedly fetch for stellar transactions (should probably happen in `vault/execution.rs` > `execute_open_requests()`) Right now it apparently only fetches once 
- [x] Make client listen to redeem events and unlock funds accordingly
- [x] Make client call the `report_stellar_transaction` of spacewalk chain that is exposed by the spacewalk pallet (eventually, the pallet has yet to be developed)
- [x] Make it possible to supply a stellar secret key to the vault via CLI arguments (is hard-coded in `vault/horizon.rs` > `fetch_latest_txs()` so far)
- [x] Remove unused code from vault. This means once everything works check the cargo logs for unused functions and remove them as well as unused files (files like `redeem.rs`, `refund.rs`, etc.). Also, the other libraries could be reduced or removed(e.g. bitcoin) but are for now part of the project because removing them is difficult because of all the interdependencies)
- [x] Test if vault works with spacewalk pallet integrated to testchain 

Check the README for instructions on how to run. 

Closes #5 and closes #6.
Edit:
Closes #26

### CODE REVIEW

**Comment by ebma** [comment]
I'd recommend changing the [name of the spacewalk pallet in the runtime](https://github.com/pendulum-chain/spacewalk/blob/af693c95123fc6c521977fa564c47e961dff90f4/testchain/runtime/src/lib.rs#L290) from `SpacewalkModule` to just `Spacewalk` because it's shorter. And the code in the PR here also expects it to be called just `Spacewalk` (this is relevant because the name is contained in the metadata fetched by subxt and has to match). 

The code that probably has to be adjusted to other types is in the `listen_for_redeem_requests` function of `vault/src/horizon.rs` and the implementation of the SpacewalkPallet trait in `runtime/src/rpc.rs`. 

After integrating the spacewalk pallet to the testchain, run the chain and use subxt (the commands are in the README of this PR) to generate a `metadata-standalone.scale` file. This file then contains the types of the event and extrinsic exposed by the pallet so after that one can change the code accordingly.

**Comment by ebma** [comment]
The testchain was replaced with a modified version of the interbtc-standalone chain to get rid of a subxt error complaining about invalid tx conversion and a codec error which I was not able to resolve otherwise. The origin was probably some type issue between what the client expected and what the chain runtime had but what exactly remains unclear. 

Changing the used keypair in the vault runtime is not feasible because dependencies that are used rely on an `AccountKeyring` which defaults to the sr25519 keypairs and not ed25519. This is why in the runtime I added both to the list of endowed accounts. At least the sr25519 identities of Alice, Bob and Charlie have to be endowed because they are used in the vaults' integration test.

**Comment by ashneverdawn** [comment]
can you resolve the merge conflicts? probably has to do with changes to the testchain. thanks!

**Comment by ashneverdawn** [comment]
Accounts I create can't make transactions because of `Inability to pay some fees` error, even though I transferred plenty of PEN tokens to them. I was able to do it with Alice but that's probably because she's sudo. What is being used to pay for fees?

**Comment by ashneverdawn** [comment]
> Alice was able to call the redeem extrinsic successfully, and the vault client did pick up the event, but I couldn't see the transaction reflected on the stellar testnet in the solar wallet.

Just noticed the error:
```
Apr 22 11:39:12.146 ERROR vault::redeem: Withdrawal transaction submission failed: HttpPostError
Apr 22 11:39:12.147 ERROR vault::redeem: Failed to process redeem request: Failed to post http request
```
Any idea why it could have failed?

**Comment by gonzamontiel** [comment]
> @ashneverdawn the errors you see appear because the redeem integration test shuts down after receiving a `RedeemEvent` although the client in the background is still trying to execute the withdrawal to stellar. So the error you see in the logs doesn't mean something goes wrong, it's just that the task in the background is terminated early. And you should also see that after running the tests it actually passes and does not fail although you see the error. I was also confused at first thinking something was wrong but I added information about this to the [README](https://github.com/pendulum-chain/spacewalk/blob/5b522d1ed32908061bf02e21f65bc22a053530d6/clients/README.md). I would argue that it is fine as it is because the withdrawal is tested in a unit test so that basically the whole flow is covered with integration + unit tests. It is difficult to test the Stellar withdrawal itself in the integration test that's why it just checks if the extrinsic call is successful and the proper event is emitted.
> 
> Regarding the other problem with `Inability to pay some fees` I also noticed this but you can work around it with [step 4 ](https://www.notion.so/satoshipay/Testing-the-Vault-and-Spacewalk-9bbdb038fbf344acaccb8aa97ca8e81a?d=f68d017a704b4be186439ed44d322569#adf18ba1d28347299da4275ac59ad10c) of this guide. There seem to somehow be multiple currencies but the one that is shown as native in the polkadot.js explorer is not the one used for sending the fees. I'll look into that again.

I realize we're using the `CurrencyId::Native` to pay fees instead of the pallet `Balances`, we shoudl revert that on the testchain.
@ne-manya can you work on that, test it again, and push it back here?

**Review by gonzamontiel — APPROVED** [review]
I left some comments, overall I think it can be approved and thus changes are optional.

File: clients/service/src/lib.rs [hunk 1]
```diff
@@ -1,8 +1,8 @@
 use async_trait::async_trait;
 use futures::{future::Either, Future, FutureExt};
 use runtime::{
-    cli::ConnectionOpts as ParachainConfig, Error as RuntimeError,
-    SpacewalkParachain as BtcParachain, SpacewalkSigner, Signer, Ss58Codec,
+    cli::ConnectionOpts as ParachainConfig, Error as RuntimeError, Signer, SpacewalkParachain as BtcParachain,
```
[the remark this answers is not in the document]
Comment (ebma) [reply clients/service/src/lib.rs]: Was not intended

File: clients/vault/src/deposit.rs [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+            .map_err(|_| Error::HttpFetchingError)?;
+
+        Ok(response)
+    }
+}
+
+struct HorizonFetcher<P: SpacewalkPallet, C: HorizonClient> {
+    parachain_rpc: P,
+    client: C,
+    escrow_secret_key: String,
```
[the remark this answers is not in the document]
Comment (ebma) [reply clients/vault/src/deposit.rs]: You're right, I changed it to vault now

File: clients/vault/src/redeem.rs [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+
+async fn execute_withdrawal(
+    escrow_secret_key: &String,
+    amount: Balance,
+    currency_id: CurrencyId,
+    destination_stellar_address: PublicKey,
+) -> Result<(), Error> {
+    let asset = CurrencyConversion::lookup(currency_id)?;
+
+    let escrow_keypair: SecretKey = SecretKey::from_encoding(escrow_secret_key).unwrap();
```
[the remark this answers is not in the document]
Comment (ebma) [reply clients/vault/src/redeem.rs]: I improved the error handling

File: clients/vault/src/redeem.rs [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+                        .unwrap()
+                    );
+                    return;
+                }
+
+                // Spawn a new task so that we handle these events concurrently
+                spawn_cancelable(shutdown_tx.subscribe(), async move {
+                    tracing::info!("Executing redeem {:?}", event);
+
+                    let currency_id = StringCurrencyConversion::convert((asset_code, asset_issuer)).unwrap();
```
[the remark this answers is not in the document]
Comment (ebma) [reply clients/vault/src/redeem.rs]: I improved the error handling

**Pull request merged**